File: src/language/ast.ts
export interface ProgramAst {
  kind: "Program";
  goal?: string;
  context?: ContextAst;
  flows: FlowAst[];
}

export interface ContextAst {
  kind: "Context";
  entries: ObjectEntryAst[];
}

export interface FlowAst {
  kind: "Flow";
  name: string;
  steps: StatementAst[];
}

export type StatementAst = ToolCallAst | RepairBlockAst;

export interface ToolCallAst {
  kind: "ToolCall";
  op: string;
  args: CallArgAst[];
  assignTo?: string;
}

export interface RepairBlockAst {
  kind: "RepairBlock";
  target: string;
  report: string;
  max: number;
  steps: StatementAst[];
}

export interface CallArgAst {
  name?: string;
  value: ExpressionAst;
}

export type ExpressionAst =
  | StringLiteralAst
  | NumberLiteralAst
  | BooleanLiteralAst
  | ArrayLiteralAst
  | ObjectLiteralAst
  | VarRefAst
  | CtxRefAst;

export interface StringLiteralAst {
  kind: "StringLiteral";
  value: string;
}

export interface NumberLiteralAst {
  kind: "NumberLiteral";
  value: number;
}

export interface BooleanLiteralAst {
  kind: "BooleanLiteral";
  value: boolean;
}

export interface ArrayLiteralAst {
  kind: "ArrayLiteral";
  items: ExpressionAst[];
}

export interface ObjectLiteralAst {
  kind: "ObjectLiteral";
  entries: ObjectEntryAst[];
}

export interface ObjectEntryAst {
  key: string;
  value: ExpressionAst;
}

export interface VarRefAst {
  kind: "VarRef";
  name: string;
}

export interface CtxRefAst {
  kind: "CtxRef";
  path: string[];
}

File: src/language/ast.rs
use serde::Serialize;
use serde_json::Number;

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct ProgramAst {
    pub goal: Option<String>,
    pub context: Option<ContextAst>,
    pub flows: Vec<FlowAst>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct ContextAst {
    pub entries: Vec<ObjectEntryAst>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct FlowAst {
    pub name: String,
    pub steps: Vec<StatementAst>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub enum StatementAst {
    ToolCall(ToolCallAst),
    RepairBlock(RepairBlockAst),
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct ToolCallAst {
    pub op: String,
    pub args: Vec<CallArgAst>,
    #[serde(rename = "assignTo", skip_serializing_if = "Option::is_none")]
    pub assign_to: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct RepairBlockAst {
    pub target: String,
    pub report: String,
    pub max: usize,
    pub steps: Vec<StatementAst>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct CallArgAst {
    pub name: Option<String>,
    pub value: ExpressionAst,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub enum ExpressionAst {
    String(String),
    Number(Number),
    Boolean(bool),
    Array(Vec<ExpressionAst>),
    Object(Vec<ObjectEntryAst>),
    VarRef(String),
    CtxRef(Vec<String>),
}

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct ObjectEntryAst {
    pub key: String,
    pub value: ExpressionAst,
}
